mypy errors with subprocess output types

When running mypy these type errors were shown around subprocess output and regex types. TODO: fixme (we silenced the errors for now)
```
src/licensemanager2/workload_managers/slurm/cmd_utils.py:32: error: Incompatible types in assignment (expression has type "str", variable has type "bytes")
src/licensemanager2/workload_managers/slurm/cmd_utils.py:42: error: Value of type variable "AnyStr" of "search" cannot be "object"
src/licensemanager2/workload_managers/slurm/cmd_utils.py:43: error: Item "None" of "Optional[Match[object]]" has no attribute "group"
src/licensemanager2/workload_managers/slurm/cmd_utils.py:43: error: Item "object" of "Union[object, Any]" has no attribute "split"
```

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -79,7 +79,7 @@ dev = [
     "uvicorn",
     "pydantic",
     "alembic",
-    ]
+]
 
 unit = [
     "aiosqlite",
@@ -96,7 +96,7 @@ unit = [
     "sqlalchemy-stubs",
     "pydantic",
     "virtualenv",
-    ]
+]
 
 mypy = [
     "mypy",
@@ -106,7 +106,12 @@ mypy = [
     "types-Jinja2",
     "types-jwt",
     "types-click",
-    ]
+    "sqlalchemy-stubs",
+    "httpx",
+    "pytest",
+    "respx",
+    "aiosqlite",
+]
 
 
 

--- a/src/licensemanager2/test/agent/test_slurmctld_prolog.py
+++ b/src/licensemanager2/test/agent/test_slurmctld_prolog.py
@@ -58,7 +58,7 @@ async def test_get_licenses_for_job_good(
         JobId=2 JobName=sleep
            UserId=root(0) GroupId=root(0) MCS_label=N/A
            Priority=4294901759 Nice=0 Account=root QOS=normal
-           JobState=CANCELLED Reason=Nodes_required_for_job_are_DOWN,_DRAINED_or_reserved_for_jobs_in_higher_priority_partitions
+           JobState=CANCELLED Reason=Nodes_required_for_job_are...
        Dependency=(null)
            Requeue=1 Restarts=0 BatchFlag=0 Reboot=0 ExitCode=0:0
            RunTime=00:00:00 TimeLimit=UNLIMITED TimeMin=N/A
@@ -99,7 +99,7 @@ async def test_get_licenses_for_job_no_licenses(
         JobId=2 JobName=sleep
            UserId=root(0) GroupId=root(0) MCS_label=N/A
            Priority=4294901759 Nice=0 Account=root QOS=normal
-           JobState=CANCELLED Reason=Nodes_required_for_job_are_DOWN,_DRAINED_or_reserved_for_jobs_in_higher_priority_partitions
+           JobState=CANCELLED Reason=Nodes_required_for_job_are...
        Dependency=(null)
            Requeue=1 Restarts=0 BatchFlag=0 Reboot=0 ExitCode=0:0
            RunTime=00:00:00 TimeLimit=UNLIMITED TimeMin=N/A

--- a/src/licensemanager2/test/agent/test_slurmctld_prolog.py
+++ b/src/licensemanager2/test/agent/test_slurmctld_prolog.py
@@ -129,7 +129,7 @@ async def test_get_licenses_for_job_no_licenses(
 
 @mark.asyncio
 @mock.patch("asyncio.create_subprocess_shell")
-async def test_get_licenses_for_job_error(
+async def test_get_licenses_for_job_command_error(
     create_process_shell_mock: mock.MagicMock,
 ):
     """
@@ -140,6 +140,22 @@ async def test_get_licenses_for_job_error(
         returncode=1,
     )
     with raises(ScontrolRetrievalFailure, match="Could not get SLURM data for job id: some-job-id"):
+        await get_licenses_for_job("some-job-id")
+
+
+@mark.asyncio
+@mock.patch("asyncio.create_subprocess_shell")
+async def test_get_licenses_for_job_output_error(
+    create_process_shell_mock: mock.MagicMock,
+):
+    """
+    Do I raise an ScontrolRetrievalFailure exception if the command output is malformed?
+    """
+    create_process_shell_mock.return_value = MockProcess(
+        "This is bad output",
+        returncode=0,
+    )
+    with raises(ScontrolRetrievalFailure, match="Command output for slurm_job_id='some-job-id' was malformed"):
         await get_licenses_for_job("some-job-id")
 
 

--- a/src/licensemanager2/test/agent/test_slurmctld_prolog.py
+++ b/src/licensemanager2/test/agent/test_slurmctld_prolog.py
@@ -2,10 +2,13 @@
 Tests of Prolog
 """
 from typing import List
-from pytest import fixture, mark
+from pytest import fixture, mark, raises
 from unittest import mock
+from licensemanager2.workload_managers.slurm.common import ENCODING
 from licensemanager2.workload_managers.slurm.cmd_utils import (
+    get_licenses_for_job,
     get_required_licenses_for_job,
+    ScontrolRetrievalFailure,
 )
 
 
@@ -30,6 +33,114 @@ def scontrol_parsed_output_bad() -> List[str]:
 @fixture
 def slurm_job_id() -> str:
     return "777"
+
+
+class MockProcess:
+
+    def __init__(self, result_str: str, returncode: int = 0):
+        self.result = result_str.encode(ENCODING)
+        self.returncode = returncode
+
+    async def communicate(self, *args, **kwargs):
+        return (self.result, None)  # Must be at least a 2-tuple
+
+
+@mark.asyncio
+@mock.patch("asyncio.create_subprocess_shell")
+async def test_get_licenses_for_job_good(
+    create_process_shell_mock: mock.MagicMock,
+):
+    """
+    Do I properly parse licenses out of `scontrol show` commands?
+    """
+    create_process_shell_mock.return_value = MockProcess(
+        """
+        JobId=2 JobName=sleep
+           UserId=root(0) GroupId=root(0) MCS_label=N/A
+           Priority=4294901759 Nice=0 Account=root QOS=normal
+           JobState=CANCELLED Reason=Nodes_required_for_job_are_DOWN,_DRAINED_or_reserved_for_jobs_in_higher_priority_partitions
+       Dependency=(null)
+           Requeue=1 Restarts=0 BatchFlag=0 Reboot=0 ExitCode=0:0
+           RunTime=00:00:00 TimeLimit=UNLIMITED TimeMin=N/A
+           SubmitTime=2021-07-19T20:22:45 EligibleTime=2021-07-19T20:22:45
+           AccrueTime=2021-07-19T20:22:45
+           StartTime=2021-07-19T20:22:49 EndTime=2021-07-19T20:22:49 Deadline=N/A
+           SuspendTime=None SecsPreSuspend=0 LastSchedEval=2021-07-19T20:22:46
+           Partition=osd-slurmd AllocNode:Sid=juju-fe651b-2:8232
+           ReqNodeList=(null) ExcNodeList=(null)
+           NodeList=(null)
+           NumNodes=1 NumCPUs=8 NumTasks=0 CPUs/Task=1 ReqB:S:C:T=0:0:*:*
+           TRES=cpu=8,billing=8
+           Socks/Node=* NtasksPerN:B:S:C=0:0:*:* CoreSpec=*
+           MinCPUsNode=1 MinMemoryNode=0 MinTmpDiskNode=0
+           Features=(null) DelayBoot=00:00:00
+              OverSubscribe=NO Contiguous=0 Licenses=test_feature.test_product@flexlm:10 Network=(null)
+           Command=sleep
+           WorkDir=/home/ubuntu
+           Power=
+           NtasksPerTRES:0
+        """,
+        returncode=0,
+    )
+    license_list = await get_licenses_for_job("some-job")
+    assert license_list == ["test_feature.test_product@flexlm:10"]
+
+
+@mark.asyncio
+@mock.patch("asyncio.create_subprocess_shell")
+async def test_get_licenses_for_job_no_licenses(
+    create_process_shell_mock: mock.MagicMock,
+):
+    """
+    Do I return an empty list if the regular expression fails to match any licenses
+    """
+    create_process_shell_mock.return_value = MockProcess(
+        """
+        JobId=2 JobName=sleep
+           UserId=root(0) GroupId=root(0) MCS_label=N/A
+           Priority=4294901759 Nice=0 Account=root QOS=normal
+           JobState=CANCELLED Reason=Nodes_required_for_job_are_DOWN,_DRAINED_or_reserved_for_jobs_in_higher_priority_partitions
+       Dependency=(null)
+           Requeue=1 Restarts=0 BatchFlag=0 Reboot=0 ExitCode=0:0
+           RunTime=00:00:00 TimeLimit=UNLIMITED TimeMin=N/A
+           SubmitTime=2021-07-19T20:22:45 EligibleTime=2021-07-19T20:22:45
+           AccrueTime=2021-07-19T20:22:45
+           StartTime=2021-07-19T20:22:49 EndTime=2021-07-19T20:22:49 Deadline=N/A
+           SuspendTime=None SecsPreSuspend=0 LastSchedEval=2021-07-19T20:22:46
+           Partition=osd-slurmd AllocNode:Sid=juju-fe651b-2:8232
+           ReqNodeList=(null) ExcNodeList=(null)
+           NodeList=(null)
+           NumNodes=1 NumCPUs=8 NumTasks=0 CPUs/Task=1 ReqB:S:C:T=0:0:*:*
+           TRES=cpu=8,billing=8
+           Socks/Node=* NtasksPerN:B:S:C=0:0:*:* CoreSpec=*
+           MinCPUsNode=1 MinMemoryNode=0 MinTmpDiskNode=0
+           Features=(null) DelayBoot=00:00:00
+              OverSubscribe=NO Contiguous=0 Network=(null)
+           Command=sleep
+           WorkDir=/home/ubuntu
+           Power=
+           NtasksPerTRES:0
+        """,
+        returncode=0,
+    )
+    license_list = await get_licenses_for_job("some-job")
+    assert license_list == []
+
+
+@mark.asyncio
+@mock.patch("asyncio.create_subprocess_shell")
+async def test_get_licenses_for_job_error(
+    create_process_shell_mock: mock.MagicMock,
+):
+    """
+    Do I raise an ScontrolRetrievalFailure exception if the return value is not 0?
+    """
+    create_process_shell_mock.return_value = MockProcess(
+        "",
+        returncode=1,
+    )
+    with raises(ScontrolRetrievalFailure, match="Could not get SLURM data for job id: some-job"):
+        await get_licenses_for_job("some-job")
 
 
 @mark.asyncio

--- a/src/licensemanager2/test/agent/test_slurmctld_prolog.py
+++ b/src/licensemanager2/test/agent/test_slurmctld_prolog.py
@@ -51,7 +51,7 @@ async def test_get_licenses_for_job_good(
     create_process_shell_mock: mock.MagicMock,
 ):
     """
-    Do I properly parse licenses out of `scontrol show` commands?
+    Do I properly parse licenses out of the `scontrol show` command?
     """
     create_process_shell_mock.return_value = MockProcess(
         """
@@ -82,7 +82,7 @@ async def test_get_licenses_for_job_good(
         """,
         returncode=0,
     )
-    license_list = await get_licenses_for_job("some-job")
+    license_list = await get_licenses_for_job("some-job-id")
     assert license_list == ["test_feature.test_product@flexlm:10"]
 
 
@@ -92,7 +92,7 @@ async def test_get_licenses_for_job_no_licenses(
     create_process_shell_mock: mock.MagicMock,
 ):
     """
-    Do I return an empty list if the regular expression fails to match any licenses
+    Do I return an empty list if the job doesn't specify any licenses?
     """
     create_process_shell_mock.return_value = MockProcess(
         """
@@ -115,7 +115,7 @@ async def test_get_licenses_for_job_no_licenses(
            Socks/Node=* NtasksPerN:B:S:C=0:0:*:* CoreSpec=*
            MinCPUsNode=1 MinMemoryNode=0 MinTmpDiskNode=0
            Features=(null) DelayBoot=00:00:00
-              OverSubscribe=NO Contiguous=0 Network=(null)
+              OverSubscribe=NO Contiguous=0 Licenses=(null) Network=(null)
            Command=sleep
            WorkDir=/home/ubuntu
            Power=
@@ -123,7 +123,7 @@ async def test_get_licenses_for_job_no_licenses(
         """,
         returncode=0,
     )
-    license_list = await get_licenses_for_job("some-job")
+    license_list = await get_licenses_for_job("some-job-id")
     assert license_list == []
 
 
@@ -139,8 +139,8 @@ async def test_get_licenses_for_job_error(
         "",
         returncode=1,
     )
-    with raises(ScontrolRetrievalFailure, match="Could not get SLURM data for job id: some-job"):
-        await get_licenses_for_job("some-job")
+    with raises(ScontrolRetrievalFailure, match="Could not get SLURM data for job id: some-job-id"):
+        await get_licenses_for_job("some-job-id")
 
 
 @mark.asyncio

--- a/src/licensemanager2/workload_managers/slurm/cmd_utils.py
+++ b/src/licensemanager2/workload_managers/slurm/cmd_utils.py
@@ -195,8 +195,15 @@ async def get_licenses_for_job(slurm_job_id: str) -> List:
     # Parse license information from scontrol output
     m = re.search('.* Licenses=([^ ]*).*', scontrol_out)
     if not m:
+        msg = f"Command output for {slurm_job_id=} was malformed: no 'Licenses' section found"
+        logger.error(msg)
+        raise ScontrolRetrievalFailure(msg)
+
+    match_string = m.group(1)
+    if match_string == "(null)":
         return []
-    license_list = m.group(1).split(',')
+
+    license_list = match_string.split(',')
     return license_list
 
 

--- a/src/licensemanager2/workload_managers/slurm/cmd_utils.py
+++ b/src/licensemanager2/workload_managers/slurm/cmd_utils.py
@@ -168,10 +168,6 @@ async def reconcile():
 async def get_licenses_for_job(slurm_job_id: str) -> List:
     """
     Parse the scontrol output and return licenses needed for job.
-
-    Note: "type: ignore" was used to silence mypy type errors
-
-    See github issue: https://github.com/omnivector-solutions/license-manager/issues/19
     """
 
     # Command to get license information back from slurm using the
@@ -182,11 +178,11 @@ async def get_licenses_for_job(slurm_job_id: str) -> List:
         stderr=asyncio.subprocess.STDOUT,
     )
 
-    scontrol_out, _ = await asyncio.wait_for(
+    scontrol_out_bytes, _ = await asyncio.wait_for(
         scontrol_show_lic.communicate(),
         CMD_TIMEOUT
     )
-    scontrol_out = str(scontrol_out, ENCODING)  # type: ignore
+    scontrol_out = scontrol_out_bytes.decode(ENCODING)
     logger.debug("##### scontrol out #####")
     logger.debug(scontrol_out)
 
@@ -197,8 +193,10 @@ async def get_licenses_for_job(slurm_job_id: str) -> List:
         raise ScontrolRetrievalFailure(msg)
 
     # Parse license information from scontrol output
-    m = re.search('.* Licenses=([^ ]*).*', scontrol_out)  # type: ignore
-    license_list = m.group(1).split(',')  # type: ignore
+    m = re.search('.* Licenses=([^ ]*).*', scontrol_out)
+    if not m:
+        return []
+    license_list = m.group(1).split(',')
     return license_list
 
 

